scene/gui/graph_edit.cpp

```diff
@@ -1646,6 +1646,585 @@ HBoxContainer *GraphEdit::get_zoom_hbox() {
+					Pair<int, int> ports(E->get().from_port, E->get().to_port);
+					if (port_info.has(_connection)) {
+						Pair<int, int> p_ports = port_info[_connection];
+						if (p_ports.first < ports.first) {
+							ports = p_ports;
+						}
+					}
+					port_info.set(_connection, ports);
+				}
+			}
+			upper_neighbours.set(gn->get_name(), s);
+		}
+	}
+
+	HashMap<int, Vector<StringName>> l = layering(selected_nodes, upper_neighbours);
+	crossing_minimisation(l, upper_neighbours);
+
+	Dictionary root, align, sink, shift;
+	horizontal_alignment(root, align, l, upper_neighbours, selected_nodes);
+
+	HashMap<StringName, Vector2> new_positions;
+	Vector2 default_position(FLT_MAX, FLT_MAX);
+	Dictionary inner_shift;
+	Set<StringName> block_heads;
+
+	for (const Set<StringName>::Element *E = selected_nodes.front(); E; E = E->next()) {
+		inner_shift[E->get()] = 0;
+		sink[E->get()] = E->get();
+		shift[E->get()] = FLT_MAX;
+		new_positions.set(E->get(), default_position);
+		if ((StringName)root[E->get()] == E->get()) {
+			block_heads.insert(E->get());
+		}
+	}
+
+	calculate_inner_shifts(inner_shift, root, node_names, align, block_heads, port_info);
+
+	List<Pair<StringName, Connection>> pending_blocks;
+	//call place_block() on each block head
+	for (const Set<StringName>::Element *E = selected_nodes.front(); E; E = E->next()) {
+		place_block(E->get(), gap_v, l, root, align, node_names, inner_shift, sink, shift, new_positions, pending_blocks);
+	}
+
+	//call post_process()
+
+	//Setting new positions of block heads
+	int block_number = 0;
+	for (const Set<StringName>::Element *E = block_heads.front(); E; E = E->next()) {
+		Vector2 cal_pos = new_positions[E->get()];
+		//Finding the layer it block head belongs to
+		unsigned int layer = 0;
+		for (; layer < l.size(); layer++) {
+			Vector<StringName> layer_nodes = l[layer];
+			if (layer_nodes.has(E->get()))
+				break;
+		}
+		cal_pos.x = origin.x + (real_t)(layer)*gap_h;
+		new_positions.set(E->get(), cal_pos);
+		block_number++;
+	}
+	//Now finding positions of nodes inside a block
+	for (const Set<StringName>::Element *E = block_heads.front(); E; E = E->next()) {
+		StringName u = E->get();
+		StringName prev;
+		int node_level = 0;
+		do {
+			Vector2 cal_pos;
+			cal_pos.y = new_positions[E->get()].y + (real_t)inner_shift[u];
+			cal_pos.x = new_positions[E->get()].x + (real_t)node_level * gap_h;
+			//Getting rid of horizontal overlap b/w current & previous node
+			if (prev != StringName()) {
+				real_t overlap = new_positions[prev].x + Object::cast_to<GraphNode>(node_names[prev])->get_size().x - cal_pos.x;
+				if (overlap > 0) {
+					cal_pos.x += overlap + gap_h / 5;
+				}
+			}
+
+			new_positions.set(u, cal_pos);
+			prev = u;
+			u = align[u];
+			node_level++;
+		} while (u != E->get());
+	}
+
+	//Moving nodes to new positions
+	emit_signal("begin_node_move");
+	for (const Set<StringName>::Element *E = selected_nodes.front(); E; E = E->next()) {
```

Comment:
We are able to use `for (const StringName &e : selected_nodes)` now I think. Need to check.

Reply:
If it works, then I'll change it all the possible places.